packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -167,6 +211,33 @@ export default createRule<Options, MessageId>({
         if (invalidAncestor.type === AST_NODE_TYPES.ReturnStatement) {
           // handle return statement
 
+          if (options.ignoreVoidInVoid) {
+            const functionNode = findFunction(invalidAncestor);
+
+            if (
+              targetNodeFunctionAncestorNodeHasVoidReturnType(invalidAncestor)
+            ) {
+              return;
+            }
+
+            if (functionNode != null) {
+              if (functionNode.type === AST_NODE_TYPES.FunctionDeclaration) {
+                if (functionNode.returnType) {
+                  return;
+                }
+              } else {
+                if (
+                  isValidFunctionExpressionReturnType(functionNode, {
+                    allowTypedFunctionExpressions: true,
+                  }) ||
+                  ancestorHasReturnType(functionNode)
```

Comment:
[Testing] All tests still pass without this line. Could you add something like to prove that we need this line?

```ts
type HigherOrderType = () => () => () => void;
const x: HigherOrderType = () => () => () => { return console.log() };
```

```ts
type HigherOrderType = () => () => () => void;
const x: HigherOrderType = () => () => function () { return console.log() };
```

Reply:
I needed that line for the above case, but I missed adding the test case. thank you!